src/CentrifugeRouter.sol

```diff
@@ -2,97 +2,186 @@
 pragma solidity 0.8.21;
 
 import {Auth} from "src/Auth.sol";
-import {Multicall} from "src/Multicall.sol";
+import {MathLib} from "src/libraries/MathLib.sol";
 import {SafeTransferLib} from "src/libraries/SafeTransferLib.sol";
+import {IERC20, IERC20Permit, IERC20Wrapper} from "src/interfaces/IERC20.sol";
 import {IERC7540Vault} from "src/interfaces/IERC7540.sol";
-import {IERC20} from "src/interfaces/IERC20.sol";
 import {ICentrifugeRouter} from "src/interfaces/ICentrifugeRouter.sol";
 import {IPoolManager} from "src/interfaces/IPoolManager.sol";
 
-contract CentrifugeRouter is Auth, Multicall, ICentrifugeRouter {
+interface EscrowLike {
+    function approve(address token, address spender, uint256 value) external;
+}
+
+contract CentrifugeRouter is Auth, ICentrifugeRouter {
+    address public immutable escrow;
+
     address public poolManager;
 
+    address constant UNSET_INITIATOR = address(1);
+    address internal _initiator = UNSET_INITIATOR;
+
     /// @inheritdoc ICentrifugeRouter
-    mapping(address user => mapping(address vault => uint256 amount)) public lockedRequests;
+    mapping(address controller => mapping(address vault => uint256 amount)) public lockedRequests;
 
-    constructor(address poolManager_) {
+    constructor(address escrow_, address poolManager_) {
+        escrow = escrow_;
         poolManager = poolManager_;
 
         wards[msg.sender] = 1;
         emit Rely(msg.sender);
     }
 
-    // --- Administration ---
-    /// @inheritdoc ICentrifugeRouter
-    function recoverTokens(address token, address to, uint256 amount) external auth {
-        SafeTransferLib.safeTransfer(token, to, amount);
+    modifier protected() {
+        if (_initiator == UNSET_INITIATOR) {
```

Comment:
We want to allow users to not go through multicall. In that case the `_initiator` should be set here. Otherwise it is set in `multicall`